src/rgw/rgw_rest_s3.cc

```diff
@@ -5870,3 +5874,201 @@ bool rgw::auth::s3::S3AnonymousEngine::is_applicable(
+
+  //2
+  buff[i++] = (char)strlen(header_name_str[CONTENT_TYPE]);
+  memcpy(&buff[i], header_name_str[CONTENT_TYPE], strlen(header_name_str[CONTENT_TYPE]));
+  i += strlen(header_name_str[CONTENT_TYPE]);
+  buff[i++] = (char)7;
+  ENCODE_NUMBER(buff[i], (short)strlen(header_value_str[OCTET_STREAM]), i);
+  memcpy(&buff[i], header_value_str[OCTET_STREAM], strlen(header_value_str[OCTET_STREAM]));
+  i += strlen(header_value_str[OCTET_STREAM]);
+
+  //3
+  buff[i++] = (char)strlen(header_name_str[MESSAGE_TYPE]);
+  memcpy(&buff[i], header_name_str[MESSAGE_TYPE], strlen(header_name_str[MESSAGE_TYPE]));
+  i += strlen(header_name_str[MESSAGE_TYPE]);
+  buff[i++] = (char)7;
+  ENCODE_NUMBER(buff[i], (short)strlen(header_value_str[EVENT]), i);
+  memcpy(&buff[i], header_value_str[EVENT], strlen(header_value_str[EVENT]));
+  i += strlen(header_value_str[EVENT]);
+
+  return i;
+}
+
+int RGWS3Select::create_message(const char *payload, char *buff)
+{
+
+  u_int32_t total_byte_len = 0;
+  u_int32_t preload_crc = 0;
+  u_int32_t message_crc = 0;
+  int i = 0;
+
+  u_int32_t header_len = creare_header_records(buff + 12);
+
+  total_byte_len = strlen(payload) + header_len + 16;
+
+  ENCODE_NUMBER(buff[i], total_byte_len, i);
+  ENCODE_NUMBER(buff[i], header_len, i);
+  char crc_buff[8];
+  memcpy(&crc_buff[0], &total_byte_len, 4);
+  memcpy(&crc_buff[4], &header_len, 4);
+
+  preload_crc = __crc32(0, (u_int8_t *)buff, 8);
+  ENCODE_NUMBER(buff[i], preload_crc, i);
+
+  i += header_len;
+  memcpy(&buff[i], payload, strlen(payload));
+  i += strlen(payload);
+
+  message_crc = __crc32(0, (u_int8_t *)buff, i);
+  ENCODE_NUMBER(buff[i], message_crc, i);
+
+  return i;
+}
+
+#define PAYLOAD_LINE "\n<Payload>\n<Records>\n<Payload>\n"
+#define END_PAYLOAD_LINE "\n</Payload></Records></Payload>"
+
+int RGWS3Select::run_s3select(char*query,char*input,size_t input_length,bool skip_first_line,bool skip_last_line,bool to_aggregate)
+{
+  char * buff = 0;
+  std::string res;
+  int status = 0;
+
+  csv_object s3_csv_object(s3select_syntax ,query ,input ,input_length ,skip_first_line ,skip_last_line ,to_aggregate );
+  if (s3select_syntax->get_error_description().empty() == false)
+  {
+    res.append(PAYLOAD_LINE);
+    res.append(s3select_syntax->get_error_description());
+    status = -1;
+  }
+  else
+  {
+    res.append(PAYLOAD_LINE);
+    status = s3_csv_object.run_s3select_on_object(res);
+    if(status<0)
+    {
+      res.append(s3_csv_object.get_error_description());
+    }
+  }
+
+  if (res.size() > strlen(PAYLOAD_LINE))
+  {
+    res.append(END_PAYLOAD_LINE);
+
+    buff = (char *)malloc(res.size() + 1000);
+    int buff_len = create_message(res.c_str(), buff);
+
+    s->formatter->write_bin_data(buff, buff_len);
+    if (op_ret < 0)
+      return op_ret;
+  }
+  rgw_flush_formatter_and_reset(s, s->formatter);
+
+  if(buff) free(buff);
+
+  return status;
+}
+
+int RGWS3Select::send_response_data(bufferlist &bl, off_t ofs, off_t len)
```

Comment:
is this invoked multiple times? if not, you can remote the data members of the class and make them local

Reply:
been called per chunk (4m)